demos/interactive_face_detection_demo/main.cpp

```diff
@@ -163,57 +143,61 @@ int main(int argc, char *argv[]) {
 
-        // --------------------------- 3. Doing inference -----------------------------------------------------
-        // Starting inference & calculating performance
-        slog::info << "Start inference " << slog::endl;
-
         bool isFaceAnalyticsEnabled = ageGenderDetector.enabled() || headPoseDetector.enabled() ||
                                       emotionsDetector.enabled() || facialLandmarksDetector.enabled();
 
+        Timer timer;
         std::ostringstream out;
         size_t framesCounter = 0;
-        int delay = 1;
-        double msrate = -1;
-        cv::Mat prev_frame, next_frame;
+        double msrate = FLAGS_fps > 0 ? 1000.0 / FLAGS_fps : 0.0;
         std::list<Face::Ptr> faces;
         size_t id = 0;
 
-        if (FLAGS_fps > 0) {
-            msrate = 1000.f / FLAGS_fps;
+        std::unique_ptr<ImagesCapture> cap = openImagesCapture(FLAGS_i, FLAGS_loop);
+        cv::Mat frame = cap->read();
+        if (!frame.data) {
+            throw std::runtime_error("Can't read an image from the input");
+        }
+
+        const cv::Point THROUGHPUT_METRIC_POSITION{10, 45};
+        Presenter presenter(FLAGS_u, THROUGHPUT_METRIC_POSITION.y + 15, {frame.cols / 4, 60});
+
+        Visualizer visualizer{frame.size()};
+        if (!FLAGS_no_show_emotion_bar && emotionsDetector.enabled()) {
+                visualizer.enableEmotionBar(emotionsDetector.emotionsVec);
         }
 
-        Visualizer::Ptr visualizer;
-        if (!FLAGS_no_show || !FLAGS_o.empty()) {
-            visualizer = std::make_shared<Visualizer>(cv::Size(width, height));
-            if (!FLAGS_no_show_emotion_bar && emotionsDetector.enabled()) {
-                visualizer->enableEmotionBar(emotionsDetector.emotionsVec);
+        cv::VideoWriter videoWriter;
+        if (!FLAGS_o.empty()) {
+            videoWriter.open(FLAGS_o, cv::VideoWriter::fourcc('I', 'Y', 'U', 'V'),
+                !FLAGS_no_show && FLAGS_fps > 0.0 ? FLAGS_fps : cap->fps(), frame.size());
+            if (!videoWriter.isOpened()) {
+                throw std::runtime_error("Can't open a video to write");
             }
         }
 
         // Detecting all faces on the first frame and reading the next one
         faceDetector.enqueue(frame);
         faceDetector.submitRequest();
 
-        prev_frame = frame.clone();
-
-        // Reading the next frame
-        bool frameReadStatus = cap.read(frame);
+        cv::Mat next_frame = cap->read();
 
         std::cout << "To close the application, press 'CTRL+C' here";
         if (!FLAGS_no_show) {
             std::cout << " or switch to the output window and press Q or Esc";
         }
         std::cout << std::endl;
 
-        const cv::Point THROUGHPUT_METRIC_POSITION{10, 45};
-
-        cv::Size graphSize{static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH) / 4), 60};
-        Presenter presenter(FLAGS_u, THROUGHPUT_METRIC_POSITION.y + 15, graphSize);
-
-        while (true) {
+        bool isLastFrame = false;
```

Comment:
Seems like this variable isn't necessary anymore - you could just replace the loop with `while (frame.data)`.

Reply:
Updated